### DISCUSSION TITLE
Bun.file().json() does not parse correctly

### DISCUSSION THREAD

**Discussion by PatrickMatthiesen** [oven-sh/bun#5018] [Q&A] [answered]
### What version of Bun is running?

1.0.0+afcbed218c425872648b7c0aa2ce331ad3b0d612

### What platform is your computer?

Linux 5.15.90.1-microsoft-standard-WSL2 x86_64 x86_64

### What steps can reproduce the bug?

`test.ts`
```ts
const file = Bun.file('some.json');
const bunJson = await file.json<Map<string, string[]>>();

const jsonText = await file.text();
const json = new Map<string, string[]>(JSON.parse(jsonText));

console.log(typeof json); // object
console.log(json.has('a')); // false

console.log(typeof bunJson); // object
console.log(bunJson.has('a')); // TypeError: bunJson.has is not a function. (In 'bunJson.has("a")', 'bunJson.has' is undefined)
```

`some.json`
```json
[
    [
        "key",
        [
            "value"
        ]
    ]
]
```

### What is the expected behavior?

file.json parses the JSON object like it says it does or change the docs.

![image](https://github.com/oven-sh/bun/assets/43612965/5baad50c-ca6d-40af-88b6-2720611d1a73)

### What do you see instead?

The JSON object is not parsed correctly, so functions do not work on the object.

### Additional information

The type is being shown as a Map in VSCode, which makes it a bit more confusing.

**Comment by alekangelov** [comment]
here you're declaring the type as a map

```ts
const bunJson = await file.json<Map<string, string[]>>();
```

but here you want it to not be a map
> The type is being shown as a Map in VSCode, which makes it a bit more confusing.

this isn't rust, types are purely decorative you can't change runtime types by declaring your types

**Comment by espenja** [comment, marked as answer]
I think you are using the API and/or TypeScript incorrectly.
`Bun` can't convert your JSON object to `Map<string, string[]>` by just using a generic type argument.
The JSON object you provided is not a "Map", it's an Array. The "Map" type doesn't exist in JSON, but is a Javascript object type, and needs to be instantiated as one. You already did this with `const json = new Map<string, string[]>(JSON.parse(jsonText));`

What you are trying to do would be the same as this:
```typescript
const sameThingWithNode = JSON.parse(jsonText) as Map<string, string[]>
console.log(sameThingWithNode.has('a')) // Doesn't show any type errors in the editor

2 | console.log(sameThingWithNode.has('a'))
               ^
TypeError: sameThingWithNode.has is not a function. (In 'sameThingWithNode.has("a")', 'sameThingWithNode.has' is undefined)
```

I believe `JSON.parse` had it's generic type parameter removed for this very reason.
VSCode is showing you the type as Map because you are telling it that it is a Map.

Try this instead:
```typescript
const file = Bun.file('some.json');
const jsonText = await file.text();
const actualBunJson = await file.json()
const jsonParseJson = JSON.parse(jsonText)

const fakeBunMap = await file.json<Map<string, string[]>>();
const fakeMapFromJsonText = JSON.parse(jsonText) as Map<string, string[]>
const mapFromJsonParse = new Map<string, string[]>(jsonParseJson);
const mapFromActualBunJson = new Map<string, string[]>(actualBunJson)

console.log("typeof actualBunJson", typeof actualBunJson) // object
console.log("typeof jsonParseJson", typeof jsonParseJson) // object

console.log("typeof fakeBunMap", typeof fakeBunMap) // object
console.log("typeof fakeMapFromJsonText", typeof fakeMapFromJsonText) // object
console.log("typeof mapFromJsonParse", typeof mapFromJsonParse) // object
console.log("typeof mapFromActualBunJson", typeof mapFromActualBunJson) // object

console.log("is fakeBunMap array?", Array.isArray(fakeBunMap)) // true
console.log("is fakeMapFromJsonText array?", Array.isArray(fakeMapFromJsonText)) // true
console.log("is mapFromJsonParse array?", Array.isArray(mapFromJsonParse)) // false
console.log("is mapFromActualBunJson array?", Array.isArray(mapFromActualBunJson)) // false

console.log("mapFromJsonParse has?", mapFromJsonParse.has('a')) // false
console.log("mapFromActualBunJson has?", mapFromActualBunJson.has('a')) // false
console.log("fakeBunMap has?", fakeBunMap.has('a')) // TypeError: fakeBunMap.has is not a function. (In 'fakeBunMap.has("a")', 'fakeBunMap.has' is undefined)
console.log("fakeMapFromJsonText has?", fakeMapFromJsonText.has('a')) // TypeError: fakeMapFromJsonText.has is not a function. (In 'fakeMapFromJsonText.has("a")', 'fakeMapFromJsonText.has' is undefined)
```

What you want is this:
```typescript
const file = Bun.file('some.json');
const actualBunJson = await file.json()
const mapFromActualBunJson = new Map<string, string[]>(actualBunJson)
```

**Reply by PatrickMatthiesen** [reply]
Thanks for taking the time to make such a good and thorough explanation, I really appreciate it.

**Comment by alekangelov** [comment]
tldr: JSON spec doesn't contain maps or specs, what you have is just a bunch of nested arrays as far as JSON.parse is concerned

https://www.json.org/json-en.html
